web3-functions/index.ts

```diff
@@ -0,0 +1,293 @@
+  livePrice: BigNumber;
+  storedPrice: BigNumber;
+  storedTimestamp: number;
+}
+
+Web3Function.onRun(async (context: Web3FunctionContext) => {
+  const { userArgs, multiChainProvider } = context;
+
+  const provider = multiChainProvider.default();
+
+  const dataServiceId = userArgs.dataServiceId as string;
+  const dataFeedIdsString = userArgs.symbols as string[];
+  const dataFeedIdsBytes32 = dataFeedIdsString.map((id) =>
+    formatBytes32String(id)
+  );
+  const oracleAddress = userArgs.oracleAddress as string;
+
+  const oracle = new Contract(oracleAddress, ORACLE_ABI, provider);
+
+  let dataFeedIds = new Map<string, DataFeed>();
+  for (const id of dataFeedIdsString) {
+    dataFeedIds.set(id, {
+      symbol: id,
+      id: formatBytes32String(id),
+      livePrice: BigNumber.from(0),
+      storedPrice: BigNumber.from(0),
+      storedTimestamp: 0,
+    });
+  }
+  debugLog("Data feed ids: ", dataFeedIds);
+
+  // Wrap contract with redstone data service
+  let wrappedOracle;
+  switch (dataServiceId) {
+    case "redstone-primary-prod":
+      wrappedOracle = WrapperBuilder.wrap(oracle).usingDataService(
+        {
+          dataServiceId: "redstone-primary-prod",
+          uniqueSignersCount: 2,
+          dataFeeds: dataFeedIdsString,
+          disablePayloadsDryRun: true,
+        },
+        [
+          "https://oracle-gateway-1.a.redstone.finance",
+          "https://oracle-gateway-2.a.redstone.finance",
+        ]
+      );
+      break;
+    case "redstone-main-demo":
+      wrappedOracle = WrapperBuilder.wrap(oracle).usingDataService(
+        {
+          dataServiceId: "redstone-main-demo",
+          uniqueSignersCount: 1,
+          dataFeeds: dataFeedIdsString,
+          disablePayloadsDryRun: true,
+        },
+        [
+          "https://oracle-gateway-1.b.redstone.finance",
+          "https://d33trozg86ya9x.cloudfront.net",
+        ]
+      );
+      break;
+    default:
+      return {
+        canExec: false,
+        message: `Data service id not found: ${dataServiceId}`,
+      };
+  }
+
+  // Retrieve live prices
+  const { data: livePriceData } =
+    await wrappedOracle.populateTransaction.getLivePricesAndTimestamp(
+      dataFeedIdsBytes32
+    );
+  const txCalldataBytes = arrayify(String(livePriceData));
+  const parsingResult = redstone.RedstonePayload.parse(txCalldataBytes);
+
+  debugLog("Unsigned metadata: ", toUtf8String(parsingResult.unsignedMetadata));
+  debugLog("Data packages count: ", parsingResult.signedDataPackages.length);
+  debugLog(getDashLine());
+
+  let dataPackageIndex = 0;
+  for (const signedDataPackage of parsingResult.signedDataPackages) {
+    debugLog(getDashLine());
+    printSignedDataPackage(dataPackageIndex, signedDataPackage);
+
+    let dataFeed = dataFeedIds.get(
+      signedDataPackage.dataPackage.dataPoints[0].dataFeedId
+    );
+
+    if (
+      dataFeed != undefined &&
+      dataFeed.symbol === signedDataPackage.dataPackage.dataPoints[0].dataFeedId
```

Comment:
Optional: the second check seems redundant as `dataFeed` is fetched from the map using the same dataFeedId.

Reply:
I think we still want to check the symbol was set correctly. But maybe there should be additional logic if the symbol is incorrect @matjazv ?

Reply:
To me it really looks like that we may remove second condition.